### ISSUE TITLE
cluster.routing.allocation.same_shard.host is not a dynamic setting (?)

### ISSUE THREAD

**Issue by nellicus** [elastic/elasticsearch#22992]
**Elasticsearch version**:
2.4

Documentation 

https://www.elastic.co/guide/en/elasticsearch/reference/2.3/shards-allocation.html#_shard_allocation_settings

```
The following dynamic settings may be used to control shard allocation and recovery:

cluster.routing.allocation.same_shard.host
Allows to perform a check to prevent allocation of multiple instances of the same shard on a single host, based on host name and host address. Defaults to false, meaning that no check is performed by default. This setting only applies if multiple nodes are started on the same machine.
```

VS

https://github.com/elastic/elasticsearch/issues/17693#issuecomment-209121270

```
The reason for this is that is "cluster.routing.allocation.same_shard.host" is not a dynamic setting, it needs to be set at startup time currently.
```

not sure if the story applies to v5.x too.

also on a slightly different note it'd be great to clarify explicitly on which node types should this be set (data/master/client/ingest). 
https://github.com/elastic/elasticsearch/issues/17693#issuecomment-209182043
seems to point out the answer is: "set to all nodes belonging to the cluster regardless of their role"

**Comment by abeyad** [comment]
I don't see a reason that this setting should not be dynamic.  I think we should fix the docs and fix the issue to make it a dynamic setting.

**Comment by jpcarey** [comment]
Ran into this today. Doc needs to be fixed.

**Comment by abeyad** [comment]
PR to fix this was just merged: #23397
